linkerd/proxy/transport/src/orig_dst.rs

```diff
@@ -0,0 +1,191 @@
+use crate::{
+    addrs::*,
+    listen::{self, Bind, Bound},
+};
+use futures::prelude::*;
+use linkerd_io as io;
+use linkerd_stack::Param;
+use std::pin::Pin;
+use tokio::net::TcpStream;
+
+#[derive(Copy, Clone, Debug, Default)]
+pub struct BindWithOrigDst<B = listen::BindTcp> {
+    inner: B,
+}
+
+#[derive(Clone, Debug)]
+pub struct Addrs<A = listen::Addrs> {
+    pub inner: A,
+    pub orig_dst: OrigDstAddr,
+}
+
+// === impl Addrs ===
+
+impl<A> Param<OrigDstAddr> for Addrs<A> {
+    fn param(&self) -> OrigDstAddr {
+        self.orig_dst
+    }
+}
+
+impl<A> Param<Remote<ClientAddr>> for Addrs<A>
+where
+    A: Param<Remote<ClientAddr>>,
+{
+    fn param(&self) -> Remote<ClientAddr> {
+        self.inner.param()
+    }
+}
+
+impl<A> Param<Local<ServerAddr>> for Addrs<A>
+where
+    A: Param<Local<ServerAddr>>,
+{
+    fn param(&self) -> Local<ServerAddr> {
+        self.inner.param()
+    }
+}
```

Comment:
sad that these can't just be a `impl<A, T> Param<T> for Addrs<A> where A: Param<T>`, but i guess there's no way to convince rustc that that wouldn't conflict with `Param<OrigDstAddrs>`...